flink-runtime/src/test/java/org/apache/flink/runtime/scheduler/StopWithSavepointTerminationHandlerImplTest.java

```diff
@@ -0,0 +1,330 @@
+                            CompletableFuture<Collection<ExecutionState>>>
+                    completion) {
+        final ExecutionState expectedNonFinishedState = ExecutionState.FAILED;
+        final String expectedErrorMessage =
+                String.format(
+                        "Inconsistent execution state after stopping with savepoint. At least one execution is still in one of the following states: %s. A global fail-over is triggered to recover the job %s.",
+                        expectedNonFinishedState, JOB_ID);
+
+        final EmptyStreamStateHandle streamStateHandle = new EmptyStreamStateHandle();
+        final CompletedCheckpoint completedSavepoint =
+                createCompletedSavepoint(streamStateHandle, "savepoint-folder");
+
+        // we have to verify that the handle pointing to the savepoint's metadata is not disposed,
+        // yet
+        assertFalse(
+                "The completed savepoint must not be disposed, yet.",
+                streamStateHandle.isDisposed());
+
+        final StopWithSavepointTerminationHandlerImpl testInstance =
+                createTestInstance(
+                        throwable ->
+                                assertThat(
+                                        throwable,
+                                        FlinkMatchers.containsMessage(expectedErrorMessage)));
+
+        final CompletableFuture<CompletedCheckpoint> completedSavepointFuture =
+                new CompletableFuture<>();
+        final CompletableFuture<Collection<ExecutionState>> executionsTerminatedFuture =
+                new CompletableFuture<>();
+
+        final CompletableFuture<String> result =
+                testInstance.handlesStopWithSavepointTermination(
+                        completedSavepointFuture,
+                        executionsTerminatedFuture,
+                        ComponentMainThreadExecutorServiceAdapter.forMainThread());
+
+        completion.accept(
+                completedSavepoint,
+                completedSavepointFuture,
+                Collections.singletonList(expectedNonFinishedState),
+                executionsTerminatedFuture);
+
+        try {
+            result.get();
+            fail("An ExecutionException is expected.");
+        } catch (Throwable e) {
+            final Optional<FlinkException> actualFlinkException =
+                    ExceptionUtils.findThrowable(e, FlinkException.class);
+            assertTrue(
+                    "A FlinkException should have been thrown.", actualFlinkException.isPresent());
+            assertThat(
+                    actualFlinkException.get(),
+                    FlinkMatchers.containsMessage(expectedErrorMessage));
+        }
+
+        // the checkpoint scheduling should be enabled in case of failure
+        assertTrue("Checkpoint scheduling should be enabled.", checkpointScheduling.isEnabled());
+
+        assertTrue("The savepoint should be cleaned up.", streamStateHandle.isDisposed());
```

Comment:
How do we assert that we call the global failover?

Reply:
Good catch! 👍